service/src/main/java/uk/nhs/adaptors/gp2gp/ehr/mapper/RequestStatementExtractor.java

```diff
@@ -0,0 +1,128 @@
+            .filter(Optional::isPresent)
+            .map(Optional::get)
+            .collect(Collectors.joining(COMMA));
+    }
+
+    public static String extractDevice(Device device) {
+        return extractTextOrCoding(device.getType())
+            .orElse(StringUtils.EMPTY);
+    }
+
+    public static String extractRecipient(MessageContext messageContext, Reference reference) {
+        IIdType referenceId = reference.getReferenceElement();
+        if (referenceId.getResourceType().equals(ResourceType.Practitioner.name())) {
+            return messageContext.getInputBundleHolder()
+                .getResource(referenceId)
+                .map(resource -> (Practitioner) resource)
+                .map(value -> RECIPIENT_PRACTITIONER + extractHumanName(value.getNameFirstRep()))
+                .orElse(StringUtils.EMPTY);
+        } else if (referenceId.getResourceType().equals(ResourceType.HealthcareService.name())) {
+            return messageContext.getInputBundleHolder()
+                .getResource(referenceId)
+                .map(resource -> (HealthcareService) resource)
+                .map(value -> RECIPIENT_HEALTH_CARE_SERVICE + value.getName())
+                .orElse(StringUtils.EMPTY);
+        } else if (referenceId.getResourceType().equals(ResourceType.Organization.name())) {
+            return messageContext.getInputBundleHolder()
+                .getResource(referenceId)
+                .map(resource -> (Organization) resource)
+                .map(value -> RECIPIENT_ORG + value.getName())
+                .orElse(StringUtils.EMPTY);
+        }
+        return StringUtils.EMPTY;
+    }
+
+    public static String extractReasonCode(ReferralRequest referralRequest) {
+        var ignoreFirstReasonCode = referralRequest.getReasonCode().subList(1, referralRequest.getReasonCode().size());
+        return ignoreFirstReasonCode.stream()
+            .map(value -> extractTextOrCoding(value))
+            .filter(Optional::isPresent)
+            .map(Optional::get)
+            .collect(Collectors.joining(COMMA));
+    }
+
+    public static String extractAuthor(MessageContext messageContext, Annotation annotation) {
+        if (annotation.hasAuthorStringType() && annotation.getAuthorStringType().hasValue()) {
+            return NOTE_AUTHOR + annotation.getAuthorStringType().getValue();
+        } else if (annotation.hasAuthorReference() && annotation.getAuthorReference().hasReferenceElement()) {
+            IIdType reference = annotation.getAuthorReference().getReferenceElement();
+            if (reference.getResourceType().equals(ResourceType.RelatedPerson.name())) {
+                return messageContext.getInputBundleHolder()
+                    .getResource(reference)
+                    .map(resource -> (RelatedPerson) resource)
+                    .map(value -> NOTE_AUTHOR_RELATION + extractHumanName(value.getNameFirstRep()))
+                    .orElse(StringUtils.EMPTY);
```

Comment:
Confirm correct behaviour if we cannot resolve the reference (current impl is to leave it blank)

Reply:
Also confirm correct behaviour if the reference is anything other than RelatedPerson, Practitioner, or Patient

Reply:
Confirmed that if a resource is not resolved an exception should be thrown. 
And that if a reference is not of expected resource type then an exception should also be thrown.